### PR TITLE
ipalib/x509: Implement abstract method for Certificate.verify_directly_issued_by

### PR DESCRIPTION
**Pull request by flo-renaud**
Taken from PR #6753 from @abbra , to fix the tox issue

### CODE REVIEW

File: ipalib/x509.py [hunk 1]
```diff
@@ -405,6 +405,18 @@ def match_hostname(self, hostname):
...
         return self._cert.tbs_precertificate_bytes
 
+    # added in python-cryptography 40.0
+    def verify_directly_issued_by(self, issuer):
+        """
+        This method verifies that certificate issuer name matches the
+        issuer subject name and that the certificate is signed by the
+        issuer's private key. No other validation is performed.
+        """
+        if 'verify_directly_issued_by' in self._cert.__abstract_methods__:
```
Comment (stanislavlevin) [inline ipalib/x509.py:415]: What is `__abstract_methods__`? `__abstractmethods__`?

Comment (stanislavlevin) [reply ipalib/x509.py:415]: `__abstractmethods__` tracks unimplemented abstract methods (that have `__isabstractmethod__` set to True). This condition is wrong.

File: ipalib/x509.py [hunk 2]
```diff
@@ -405,6 +405,18 @@ def match_hostname(self, hostname):
...
+            return self._cert.verify_directly_issued_by(issuer)
+        else:
+            return None
+
```
Comment (tiran) [inline ipalib/x509.py:419]: This approach defines the additional concrete method if and only if the ABC has the new method.

```suggestion
    if hasattr(crypto_x509, "verify_directly_issued_by"):
        # added in python-cryptography 40.0
        def verify_directly_issued_by(self, issuer):
            return self._cert.verify_directly_issued_by(issuer)
```

Comment (stanislavlevin) [reply ipalib/x509.py:419]: typo:
`crypto_x509` => `crypto_x509.Certificate`.

**Comment by stanislavlevin** [comment]
@flo-renaud,
LGTM, thank you.

- checked `IPACertificate`'s `verify_directly_issued_by` against `cryptography-40.0.1`.

**Comment by flo-renaud** [comment]
Azure is now green, ready for final review.

**Comment by stanislavlevin** [comment]
ACK here.

**Comment by flo-renaud** [comment]
master:

* 3721bca657264a928926706c75ab0d2a7593845f ipalib/x509: Implement abstract method Certificate.verify_directly_issued_by

**Pull request closed without merging**